xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,246 @@
+    ) {
+        spawn RLEDec<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:2;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount  = bits[TEST_COUNT_WIDTH];
+type TestDecInData  = DecInData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+type TestDecOutData = DecOutData<TEST_SYMBOL_WIDTH>;
+
+
+//const TEST_SEND_LOOKUP = TestDecInData[10]:[
+//    TestDecInData {symbol: TestSymbol:0x1, count: TestCount:0x1, last:true},
+//    TestDecInData {symbol: TestSymbol:0x2, count: TestCount:0x1, last:true},
+//];
+//
+//const TEST_RECV_LOOKUP = TestDecOutData[20]:[
+//    TestDecOutData {symbol: TestSymbol:0x1, last: true},
+//    TestDecOutData {symbol: TestSymbol:0x2, last: true},
+//];
+
+// main proc used to test the RLE decoder
+#[test_proc]
+proc RLEDecTightTester {
+    terminator: chan<bool> out;            // test termination request
+    dec_input_s: chan<TestDecInData> out;
+    dec_output_r: chan<TestDecOutData> in;
+
+    init {()}
+
+    config(terminator: chan<bool> out) {
+        let (dec_input_s, dec_input_r)   = chan<TestDecInData>;
+        let (dec_output_s, dec_output_r) = chan<TestDecOutData>;
+
+        spawn RLEDec<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(dec_input_r, dec_output_s);
+        (terminator, dec_input_s, dec_output_r)
+    }
+
+    next(tok: token, state: ()) {
+
+        // Simple transaction without repeats
+        let trans_send: (TestSymbol, TestCount)[2] =
+            [(TestSymbol:0xA, TestCount:0x3), (TestSymbol:0xB, TestCount:0x1)];
+        let tok = for ((counter, (symbol_in, count_in)), tok): ((u32, (TestSymbol, TestCount)) , token) in enumerate(trans_send) {
+            let _last = counter == (array_size(trans_send) - u32:1);
+            let data_in = TestDecInData{symbol: symbol_in, count: count_in, last: _last};
+            let tok = send(tok, dec_input_s, data_in);
+            let _ = trace_fmt!("Sent {} transactions, symbol: 0x{:x}, count:{}, last: {}",
+                counter, data_in.symbol, data_in.count, data_in.last);
+            (tok)
+        }(tok);
+        let trans_recv: TestSymbol[4] = [
+            TestSymbol: 0xA, TestSymbol: 0xA, TestSymbol: 0xA, TestSymbol: 0xB
+        ];
+        let tok = for ((counter, symbol_out), tok): ((u32, TestSymbol) , token) in enumerate(trans_recv) {
+            let _last = counter == (array_size(trans_recv) - u32:1);
+            let data_out = TestDecOutData{symbol: symbol_out, last: _last};
+            let (tok, dec_output) = recv(tok, dec_output_r);
+            let _ = trace_fmt!(
+                "Received {} transactions, symbol: 0x{:x}, last: {}",
+                counter, dec_output.symbol, dec_output.last
+            );
+            let _ = assert_eq(dec_output, data_out);
+            (tok)
+        }(tok);
+
+        // Transaction with repeating symbols
```

Comment:
is the different thing we're testing here the symbol `0xC` being chunked across two packet? shouldn't that meant that we set `last` on the first one (as we don't overflow the only time it can happen is with two transactions)

Reply:
same commment here, is it valid in the protocol to heave multiple consecutive `(symbol, count)` packet with the same `symbol` without having `last` being set between them?

Reply:
It's possible to receive multiple `(symbol, count)` pairs with the same `symbol` without `last`being set in the first one.
This can happen when either encoder's counter has overflown or some preceding stage has decided to send such packets. Decoder must be able to handle such transaction.